Adds documentation about value objects (#256)
PrestaShop/docs: src/content/1.7/development/architecture/domain/value_objects.md

---
title: Value Objects
weight: 20
---

# Value Objects

> PrestaShop is using Value Objects in its codebase, to help you understand better what they are, this article aims to explain what Value object is and how to use it.

Reviewer: ```suggestion
PrestaShop is using Value Objects in its codebase. To help you understand better what they are, this article aims to explain what Value object is and how to use it.
```

## What is Value Object?

> A Value Object is small immutable object whos equality is not based on identity, this means that two Value Objects are equal when they have the same value, but not necessarily being the same object.

Reviewer: ```suggestion
[Value Objects](https://martinfowler.com/bliki/ValueObject.html) (VO for short) are small, simple, immutable objects, like Money or a date range, whose equality is not based on their identity. This means that any two given instances of a same VO are considered equal when they contain the same values.

Value Objects are often referred to as Value Types as well. You can consider a Value Object like a data _type_, whose objective is to provide structure, validation, and meaning.
```

Characteristics of Value Object are:

> * It does not have identity (e.g. `Money` is value object because it can be created without id like `new Money($amount)`).

Reviewer: ```suggestion
* It does not have an identity, contrary to an Entity (e.g. `Money($amount)` can be considered a VO because two instances of it can be considered equal as long as the `$amount` is equal, regardless of which bills that amount is made of; but `Bill($billNumber)` is an entity because it refers to a specific physical Bill).
```

> * It is immutable (e.g. it cannot be modified, but instead new instance of value object is created after its modification).

Reviewer: ```suggestion
* It is immutable (e.g. it cannot be modified; any modification on it yields a different instance).
```

* It is self validating (e.g. value object cannot be created with invalid values, meaning that attempt to create `new Email('not an email')` would throw an exception).
* It is interchangeable (e.g. `$a` and `$b` can replace one another without any side effects if they both are created like `$a = new Money(100)` and `$b = new Money(100)`).

## Examples of Value Objects in PrestaShop

As an example, we can take a look at `PrestaShop\PrestaShop\Core\Domain\Currency\ValueObject\ExchangeRate` value object. `ExchangeRate` is used to pass currency exchange rate between different parts of system.

```php
use PrestaShop\PrestaShop\Core\Domain\Currency\ValueObject\ExchangeRate;

// ExchangeRate has VO has rules to protect.
// In this case, ExchangeRate validates that it's value cannot be 0 or less.
// If we were to create ExchangeRate with value of -1,

> // then CurrencyConstraintException would be thrown.

Reviewer: I think it's better to leave long texts outside the code section

Author: Ummm, I see that Symfony docs are using comments to explain code.

In addition, I believe that many developers just skip reading documentation, but dive straight into code example, so even if it duplicates content outside the code, it's not necessarily a bad thing imo.

But I can remove it if you wish. :)

Reviewer 2: Agreed with @sarjon :roll_eyes:
